server/text_generation_server/models/causal_lm.py

```diff
@@ -1041,23 +1039,28 @@ def generate_token(self, batches: List[CausalLMBatch]) -> Tuple[List[Generation]
         return generations, batch if not stopped else None
 
     def warmup(self, batches: List[CausalLMBatch]) -> None:
-        self.shifting_warmup()
-
         if len(batches) < 2:
             return
 
         # prefill
-        _, prefill_batch = self.generate_token([batches[0]])
+        _, prefill_batch = self.generate_token([batches.pop(0)])
         # decode
         _, decode_batch = self.generate_token([prefill_batch])
+        # shifts
+        self.shifting_warmup(decode_batch)
```

Comment:
I think you could also add warmup for shifting prefill_batch . In case of concat we usually pad the newly prefilled batch to the right

Reply:
You could only limit it to chunk > 0, because we never shift left in concat

Reply:
I tried this, but after prefill kv_cache is not accessible yet (batch.past_key_values == None). It is updated during next generate_token, and right after that prefill batch is destroyed (concatenated with decode batch).